### PR TITLE
[ADP-3297] Support restoration mode in CLI

### PR DESCRIPTION
**Pull request by paolino**
- [x] Add support for restoration mode options in the CLI interface

ADP-3297

### CODE REVIEW

**Review by paweljakubas — CHANGES_REQUESTED** [review]
please add unit test here https://github.com/cardano-foundation/cardano-wallet/blob/master/lib/unit/test/unit/Cardano/CLISpec.hs

File: lib/wallet/api/http/Cardano/CLI.hs [hunk 1]
```diff
@@ -655,7 +661,40 @@ cmdWalletCreateFromMnemonic mkClient =
...
+    fromTipCmd =
+        command "from-tip"
+            $ info
+                (f <*> pure RestoreFromTip)
+            $ progDesc "Restore the wallet from the tip."
+    fromCheckpointCmd =
+        command "from-checkpoint"
+            $ info
+                (f <*> fromCheckpointP)
+            $ progDesc "Restore the wallet from a specific checkpoint."
```
Comment (paweljakubas) [inline lib/wallet/api/http/Cardano/CLI.hs:689]: what about adding adding additional information here? To help user to use it, like
```
        <> progDesc "Restore the wallet from a specific checkpoint."
        <> header "Checkpoint can be specified either via block header hash or slot no"
        <> footerDoc (Just $ vsep
            [ string "Example for block header hash:"
            , indent 2 $ bold $ string $ progName<>" ...... "
            , string "Example for slot number:"
            , indent 2 $ bold $ string $ progName<>" ...... "
              ])
```
and in this examples please try to provide real header hash that was there a week ago and respective slot No.

**Review by paweljakubas — CHANGES_REQUESTED** [review]

File: lib/wallet/api/http/Cardano/CLI.hs [hunk 2]
```diff
@@ -655,7 +664,56 @@ cmdWalletCreateFromMnemonic mkClient =
...
+            <> header "Restoration from checkpoint (partial restoration)"
+            <> footerDoc (Just fromCheckpointCmdFooter)
+
+    fromCheckpointCmdFooter = vsep [
+        "This is the way to restore the wallet from a specific checkpoint."
+        , "It will create a partial view of the wallet if the wallet has \
+            \transactions before the checkpoint!"
+        , "You will need a blockheader hash and a slot number to restore from."
+        , "Example:"
```
Comment (paweljakubas) [inline lib/wallet/api/http/Cardano/CLI.hs:698]:
```    , "Example:"
        , indent 2 "cardano-wallet wallet create from-recovery-passphrase \
            \from-checkpoint \
            \--block-header-hash 95af5a1da7d43571d030c92ba82728d69c86c82007cff6ee3c8b2442ab148723 \
            \'My wallet'"
        , "Example:"
        , indent 2 "cardano-wallet wallet create from-recovery-passphrase \
            \from-checkpoint \
            \--slot-no 118332737 \
            \'My wallet'"
```
You specify EITHER block header hash OR slot number. Not both, right?

**Review by paweljakubas — APPROVED** [review]
lgtm modulo doc comment

File: lib/wallet/api/http/Cardano/CLI.hs (hunk 2 above)
Comment (paolino) [reply lib/wallet/api/http/Cardano/CLI.hs:698]: Only both (sadly). This is a limitation coming from 

1. We are not storing block information in the DB
2. The node-to-client protocol needs both to start the chain-sync

Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/CLI.hs:698]: ok, now understand the constraint. Thanks!

**Pull request merged**